### ISSUE TITLE
[clang-tidy] False positive bugprone-std-namespace-modification with templated user-defined types

### ISSUE THREAD

**Issue by carlosgalvezp** [llvm/llvm-project#183752]
Example:

```cpp
#include <functional>

struct Foo
{};

template <typename T>
struct Bar
{};

template <>
struct std::hash<Foo> 
{};


template <typename T>
struct std::hash<Bar<T>> 
{};
```

Gives:
```
source>:16:13: warning: modification of 'std' namespace can result in undefined behavior [bugprone-std-namespace-modification]
   16 | struct std::hash<Bar<T>> 

```
[Godbolt](https://godbolt.org/z/c3EPvWxje)
